bindings/cpp/newapi/session.cpp

```diff
@@ -844,4 +845,221 @@ async_iterator_result session::server_send(const std::vector<key> &keys, uint64_
+				   const dnet_addr &addr)
+	: m_session(orig_sess)
+	, m_handler(result)
+	, m_last_error(0)
+	, m_addr(addr)
+	, m_log(orig_sess.get_logger()) {
+	}
+
+	void start(const transport_control &control, const dnet_bulk_read_request &request) {
+		DNET_LOG_NOTICE(m_log, "{}: started: address: {}, flags: {}, read_flags: {}, num_keys: {}",
+				dnet_cmd_string(control.get_native().cmd), dnet_addr_string(&m_addr),
+				dnet_flags_dump_ioflags(request.ioflags), request.read_flags, request.keys.size());
+
+		auto rr = send_to_single_state(m_session, control);
+		m_handler.set_total(rr.total());
+
+		m_keys.assign(request.keys.begin(), request.keys.end());
+		std::sort(m_keys.begin(), m_keys.end(), key_compare);
+		m_key_responses.resize(m_keys.size(), false);
+
+		rr.connect(
+			std::bind(&bulk_read_handler::process, shared_from_this(), std::placeholders::_1),
+			std::bind(&bulk_read_handler::complete, shared_from_this(), std::placeholders::_1)
+		);
+	}
+
+private:
+	void process(const ioremap::elliptics::callback_result_entry &entry) {
+		auto cmd = entry.command();
+
+		const auto &resp = callback_cast<read_result_entry>(entry);
+		if (!resp.is_valid()) {
+			DNET_LOG_ERROR(m_log, "{}: {}: process: invalid response, status: {}",
+				       dnet_dump_id(&cmd->id), dnet_cmd_string(cmd->cmd), cmd->status);
+			return;
+		}
+
+		auto it = std::lower_bound(m_keys.begin(), m_keys.end(), cmd->id, key_compare);
+
+		bool found = false;
+		if (it != m_keys.end()) {
+			const auto index = std::distance(m_keys.begin(), it);
+			if (dnet_id_cmp(&cmd->id, &m_keys[index]) == 0) {
+				m_handler.process(resp);
+
+				m_key_responses[index] = true;
+				found = true;
+			}
+		}
+
+		if (!found) {
+			DNET_LOG_ERROR(m_log, "{}: {}: process: unknown key, status: {}",
+				       dnet_dump_id(&cmd->id), dnet_cmd_string(cmd->cmd), cmd->status);
+		}
+
+		m_last_error = cmd->status;
+	}
+
+	void complete(const error_info &error) {
+		dnet_cmd cmd;
+		memset(&cmd, 0, sizeof(cmd));
+		cmd.status = error ? error.code() : m_last_error;
+		cmd.cmd = DNET_CMD_BULK_READ_NEW;
+		cmd.trace_id = m_session.get_trace_id();
+		cmd.flags = DNET_FLAGS_REPLY | DNET_FLAGS_MORE |
+			(m_session.get_trace_bit() ? DNET_FLAGS_TRACE_BIT : 0);
+
+		for (size_t i = 0; i < m_keys.size(); ++i) {
+			if (m_key_responses[i])
+				continue;
+
+			cmd.id = m_keys[i];
+			auto result_data = std::make_shared<ioremap::elliptics::callback_result_data>(&m_addr, &cmd);
+			result_data->error = error ? error :
+				create_error(m_last_error, "send_bulk_read: read failed for key: %s",
+					     dnet_dump_id(&m_keys[i]));
+			ioremap::elliptics::callback_result_entry entry(result_data);
+			m_handler.process(callback_cast<read_result_entry>(entry));
+		}
+
+		m_handler.complete(error);
+
+		DNET_LOG_NOTICE(m_log, "{}: finished: address: {}",
+				dnet_cmd_string(DNET_CMD_BULK_READ_NEW), dnet_addr_string(&m_addr));
+	}
+
+	inline static bool key_compare(const dnet_id &lhs, const dnet_id &rhs) {
```

Comment:
`key_compare` is not required since you have added `bool operator <(const dnet_id &lhs, const dnet_id &rhs)`.